src/hostapi/pulseaudio/pa_linux_pulseaudio.c

```diff
@@ -0,0 +1,1397 @@
+
+        /* unless alternate device specification is supported, reject the use of
+         * paUseHostApiSpecificDeviceSpecification
+         */
+        if( inputParameters->device == paUseHostApiSpecificDeviceSpecification )
+        {
+            result = paInvalidDevice;
+            goto openstream_error;
+        }
+
+        /* check that input device can support inputChannelCount */
+        if( inputChannelCount >
+            hostApi->deviceInfos[inputParameters->device]->maxInputChannels )
+        {
+            result = paInvalidChannelCount;
+            goto openstream_error;
+        }
+
+        /* validate inputStreamInfo */
+        if( inputParameters->hostApiSpecificStreamInfo )
+        {
+            /* this implementation doesn't use custom stream info */
+            result = paIncompatibleHostApiSpecificStreamInfo;
+            goto openstream_error;
+        }
+
+        hostInputSampleFormat =
+            PaUtil_SelectClosestAvailableFormat( inputSampleFormat,
+                                                 inputSampleFormat );
+
+        stream->inputFrameSize = Pa_GetSampleSize( inputSampleFormat ) * inputChannelCount;
+
+        result = PaPulseAudio_ConvertPortaudioFormatToPaPulseAudio_(
+            hostInputSampleFormat,
+            &stream->inSampleSpec );
+
+        if( result != paNoError )
+        {
+            goto openstream_error;
+        }
+
+        stream->inSampleSpec.rate = sampleRate;
+        stream->inSampleSpec.channels = inputChannelCount;
+        stream->latency = inputParameters->suggestedLatency;
+
+        if( !pa_sample_spec_valid(&stream->inSampleSpec) )
+        {
+            PA_DEBUG( ("Portaudio %s: Invalid input audio spec!\n",
+                      __FUNCTION__) );
+            result = paUnanticipatedHostError; /* should have been caught above */
+            goto openstream_error;
+        }
+
+        stream->inStream =
+            pa_stream_new( l_ptrPulseAudioHostApi->context,
+                           stream->sourceStreamName,
+                           &stream->inSampleSpec,
+                           NULL );
+
+        if( stream->inStream != NULL )
+        {
+            pa_stream_set_state_callback( stream->inStream,
+                                          PaPulseAudio_StreamStateCb,
+                                          stream);
+            pa_stream_set_started_callback( stream->inStream,
+                                            PaPulseAudio_StreamStartedCb,
+                                            stream );
+        }
+        else
+        {
+            PA_DEBUG( ("Portaudio %s: Can't alloc input stream!\n",
+                       __FUNCTION__) );
+        }
+
+        stream->inDevice = inputParameters->device;
+
+        result = PaPulseAudio_BlockingInitRingBuffer( stream,
+                                                      &stream->inputRing,
+                                                      (4096 * 32) );
```

Comment:
What is the reason for these numbers? Can it become: 
```suggestion
                                                      stream->inputFrameSize * framesPerBuffer * inputChannelCount );
```